src/main/java/io/supertokens/bulkimport/BulkImportUserUtils.java

```diff
@@ -0,0 +1,167 @@
+ *    obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *    License for the specific language governing permissions and limitations
+ *    under the License.
+ */
+
+package io.supertokens.bulkimport;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+import io.supertokens.bulkimport.exceptions.InvalidBulkImportDataException;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser.LoginMethod;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser.LoginMethod.EmailPasswordLoginMethod;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser.LoginMethod.ThirdPartyLoginMethod;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser.LoginMethod.PasswordlessLoginMethod;
+import io.supertokens.pluginInterface.bulkimport.BulkImportUser.TotpDevice;
+import io.supertokens.pluginInterface.utils.JsonValidatorUtils.ValueType;
+import io.supertokens.utils.Utils;
+
+import static io.supertokens.pluginInterface.utils.JsonValidatorUtils.parseAndValidateField;
+import static io.supertokens.pluginInterface.utils.JsonValidatorUtils.validateJsonFieldType;
+
+public class BulkImportUserUtils {
+    public static BulkImportUser createBulkImportUserFromJSON(JsonObject userData, String id) throws InvalidBulkImportDataException {
+        List<String> errors = new ArrayList<>();
+
+        String externalUserId = parseAndValidateField(userData, "externalUserId", ValueType.STRING, false, String.class, errors, ".");
+        JsonObject userMetadata = parseAndValidateField(userData, "userMetadata", ValueType.OBJECT, false, JsonObject.class, errors, ".");
+        List<String>  userRoles = getParsedUserRoles(userData, errors);
+        List<TotpDevice> totpDevices = getParsedTotpDevices(userData, errors);
+        List<LoginMethod> loginMethods = getParsedLoginMethods(userData, errors);
+
+        if (!errors.isEmpty()) {
+            throw new InvalidBulkImportDataException(errors);
+        }
+        return new BulkImportUser(id, externalUserId, userMetadata, userRoles, totpDevices, loginMethods);
+    } 
+
+    private static List<String> getParsedUserRoles(JsonObject userData, List<String> errors) {
+        JsonArray jsonUserRoles = parseAndValidateField(userData, "roles", ValueType.ARRAY_OF_STRING,
+                false,
+                JsonArray.class, errors, ".");
+
+        if (jsonUserRoles == null) {
+            return null;
+        }
+
+        List<String> userRoles = new ArrayList<>();
+        jsonUserRoles.forEach(role -> userRoles.add(role.getAsString().trim()));
```

Comment:
dont we also make roles lowecase? What what if it's an empty string?

Reply:
1. We don't make role lowercase.
2. This won't be an empty string because `JsonValidatorUtils.validateJsonFieldType` doesn't consider `""` as valid string type and will return null if it finds empty string.